### DISCUSSION TITLE
Contextual substitution of a mark glyph

### DISCUSSION THREAD

**Discussion by murasu** [harfbuzz/harfbuzz#4509] [General] [answered]
I'm trying to contextually substitute an above based mark with the following GSUB code in psts:
`sub ka-tamil iMatra-tamil' pa-tamil by iMatra-tamil.conn1;`

The substitution works in CoreText but not in Harfbuzz. The image below will show and compare the outcomes. Green is CoreText.

<img width="1368" alt="Screenshot 2023-12-02 at 12 45 35 AM" src="https://github.com/harfbuzz/harfbuzz/assets/22926756/5d39c059-3c89-425e-b3ec-928b23029f7b">

I'm attaching the font as well. Is there anything I need to do different for HB to render this rule?

[MarkSubTest-Light.otf.zip](https://github.com/harfbuzz/harfbuzz/files/13530030/MarkSubTest-Light.otf.zip)

**Comment by murasu** [comment]
I did some more tests. The only way I could get the substitution work on Harfbuzz is to add an empty sustitution to the chain as below:

```
script tml2;

lookup subst_imatra {
	sub iMatra-tamil by iMatra-tamil.conn1;
} subst_imatra;

lookup do_nothing {
} do_nothing;


sub ka-tamil' lookup do_nothing iMatra-tamil' lookup subst_imatra pa-tamil;

```

This works across CoreText and Harfbuzz ... but just for two glyphs in a sequence. Still puzzled :(

**Reply by khaledhosny** [reply]
It works if you put the substitution in `calt`, so I suspect it is because `psts` feature is applied per syllable. I don’t understand why adding more input makes it work across syllable boundaries, though.

**Reply by behdad** [reply]
If anything it should be the other way around. We do match backtrack/lookahead sequences across grapheme boundaries always IIRC.

cc @jfkthame

**Reply by murasu** [reply]
Yes, the substitution works if it's in `calt` or `rclt` without more inputs.

**Reply by khaledhosny** [reply]
FWIW, DirectWrite and Uniscribe seems to not apply the substitution regardless of the feature tag used.

**Reply by behdad** [reply]
Interesting. They probably match backtrack/lookahead within syllable only.

**Comment by behdad** [comment]
Can you attach the working font as well? Thanks.

**Comment by murasu** [comment]
Here's the working font.  

<img width="1428" alt="Screenshot 2023-12-02 at 7 36 19 PM" src="https://github.com/harfbuzz/harfbuzz/assets/22926756/afdf660e-1e4e-4ad9-8130-4b3a8ce15f9c">


[MarkSubTest-Light 2.otf.zip](https://github.com/harfbuzz/harfbuzz/files/13536129/MarkSubTest-Light.2.otf.zip)

**Comment by behdad** [comment]
I see a bug...

**Comment by behdad** [comment, marked as answer]
I'm working on it in https://github.com/harfbuzz/harfbuzz/pull/4510
